File: ext/use-lib.rb
#!/usr/bin/ruby

require 'ffi'

# see https://github.com/ffi/ffi/issues/467#issuecomment-159370223

module Stdio
  extend FFI::Library
  ffi_lib FFI::Platform::LIBC
  attach_function :free, [ :pointer ], :void
end

module SUSEConnect
  extend FFI::Library
  ffi_lib './libsuseconnect.so'

  attach_function :getstatus, [:string], :pointer
end

p_out = SUSEConnect.getstatus('json')
puts p_out.get_string(0)
Stdio.free(p_out)

File: ext/use-lib.c
/*
    go build -buildmode=c-shared -o libsuseconnect.so ext/main.go
    gcc ext/use-lib.c -o use-lib -L. -lsuseconnect
    LD_LIBRARY_PATH=. ./use-lib
*/

#include <stdio.h>
#include <stdlib.h>
#include "../libsuseconnect.h"

void main(void) {
    char *format = "json";
    char *statuses;

    statuses = getstatus(format);
    printf("%s\n", statuses);
    free(statuses);
}
